container/go/pkg/compat/overrideConfig.go

```diff
@@ -0,0 +1,453 @@
+			if extractedValue, err = Stamp(value, overrideInfo.StampInfoFile); err != nil {
+				return errors.Wrapf(err, "Failed to format the string accordingly at %s", value)
+			}
+			labels[label] = extractedValue
+		}
+	}
+	if len(overrideInfo.LabelsArray) > 0 {
+		labelsMap := make(map[string]string)
+		if len(overrideInfo.ConfigFile.Config.Labels) > 0 {
+			labelsMap = overrideInfo.ConfigFile.Config.Labels
+		}
+		for k, v := range labels {
+			labelsMap[k] = v
+		}
+		overrideInfo.ConfigFile.Config.Labels = labelsMap
+	}
+
+	if len(overrideInfo.Ports) > 0 {
+		if len(overrideInfo.ConfigFile.Config.ExposedPorts) == 0 {
+			overrideInfo.ConfigFile.Config.ExposedPorts = make(map[string]struct{})
+		}
+		for _, port := range overrideInfo.Ports {
+			match, err := regexp.MatchString("[0-9]+/(tcp|udp)", port)
+			if err != nil {
+				return errors.Wrapf(err, "Failed to successfully match regex to %s", port)
+			}
+			if match {
+				// the port spec has the form 80/tcp, 1234/udp so simply use it as the key.
+				overrideInfo.ConfigFile.Config.ExposedPorts[port] = struct{}{}
+			} else {
+				// assume tcp
+				overrideInfo.ConfigFile.Config.ExposedPorts[port+"/tcp"] = struct{}{}
+			}
+		}
+	}
+
+	if len(overrideInfo.Volumes) > 0 {
+		if len(overrideInfo.ConfigFile.Config.Volumes) == 0 {
+			overrideInfo.ConfigFile.Config.Volumes = make(map[string]struct{})
+		}
+		for _, volume := range overrideInfo.Volumes {
+			overrideInfo.ConfigFile.Config.Volumes[volume] = struct{}{}
+		}
+	}
+
+	if overrideInfo.Workdir != "" {
+		stampedWorkdir, err := Stamp(overrideInfo.Workdir, overrideInfo.StampInfoFile)
+		if err != nil {
+			return errors.Wrapf(err, "Unable to stamp the working directory %s", overrideInfo.Workdir)
+		}
+		overrideInfo.ConfigFile.Config.WorkingDir = stampedWorkdir
+	}
+
+	// layerDigests are diffIDs extracted from each layer file.
+	layerDigests := []string{}
+	for _, l := range overrideInfo.Layer {
+		newLayer, err := extractValue(l)
+		if err != nil {
+			return errors.Wrap(err, "Failed to extract the contents of layer file: %v")
+		}
+		layerDigests = append(layerDigests, newLayer)
+	}
+	// diffIDs are ordered from bottom-most to top-most.
+	diffIDs := []v1.Hash{}
+	if len(overrideInfo.ConfigFile.RootFS.DiffIDs) > 0 {
+		diffIDs = overrideInfo.ConfigFile.RootFS.DiffIDs
+	}
+	if len(overrideInfo.Layer) > 0 {
+		var diffIDToAdd v1.Hash
+		for _, diffID := range layerDigests {
+			if diffID != emptySHA256Digest() {
+				diffIDToAdd = v1.Hash{Algorithm: "sha256", Hex: diffID}
+				diffIDs = append(diffIDs, diffIDToAdd)
+			}
+		}
+		overrideInfo.ConfigFile.RootFS = v1.RootFS{Type: "layers", DiffIDs: diffIDs}
+
+		// length of history is expected to match the length of diff_ids.
+		history := []v1.History{}
+		if len(overrideInfo.ConfigFile.History) > 0 {
+			history = overrideInfo.ConfigFile.History
+		}
+		var historyToAdd v1.History
+
+		for _, l := range layerDigests {
+			var createdBy = overrideInfo.CreatedByArg
+			if createdBy == "" {
```

Comment:
Based on how OverrideConfig is called, createdBy can never be empty. I would suggest making this and the check for `Author` below a validate function on `OverrideConfigOpts`. Something like

```
func (o *OverrideConfigOpts) validate() error {
   if o.CreatedBy == "" {
       return errors.New("CreatedBy was not specified")
   }
   ...
}
```

and just call
```
if err := overrideInfo.validate(); err != nil {
   return errors.Wrap(err, "unable to create a new config because the given options to override the existing image config/generate a new config failed validation")
}
```